WIP client: Handle server disconnects

charlesbaynham opened the pull request:
The second half of https://github.com/m-labs/artiq/pull/1361, aiming to handle server disconnects sensibly in the rpc clients. 

We previously discussed using similar logic to `sync_struct` to handle disconnects. `sync_struct` treats a server disconnect as a normal event and just finishes its task. It also invokes a callback, if supplied. 

However, I don't think that works here because neither the `Client` nor the `AsyncioClient` have an infinite loop running, from which we can exit on disconnect. Both objects are queried by the user code, and so both need to signal to the user that the server has closed the connection and it will never work again. I think this is what exceptions are for.

I've therefore raised `ConnectionAbortedError` when the user calls methods that interact with the server but the server has shut down. 

Remaining questions are what to do with `BestEffortClient`? This currently returns None for methods that fail due to connection errors, but a closed socket will never be resolved. Should it attempt to reopen the socket each time it's called?

charlesbaynham commented:
@sbourdeauducq would appreciate your thoughts.

sbourdeauducq commented on sipyco/pc_rpc.py:
```diff
@@ -24,6 +24,8 @@
 from sipyco.asyncio_tools import AsyncioServer as _AsyncioServer
 from sipyco.packed_exceptions import *
 
+CONNECTION_CLOSED_ERR = ConnectionAbortedError("Connection closed by the server")
```
Why this variable?

charlesbaynham replied:
Because there are two places where it needs to be raised and I didn't want to have to repeat the message.

sbourdeauducq commented on sipyco/pc_rpc.py:
```diff
@@ -603,7 +630,7 @@ async def _handle_connection_cr(self, reader, writer):
                     break
                 reply = await self._process_and_pyonize(target,
-                    pyon.decode(line.decode()))
+                                                        pyon.decode(line.decode()))
```
Please move all the style changes into a separate PR.

charlesbaynham replied:
Done, see #12

sbourdeauducq commented:
> This currently returns None for methods that fail due to connection errors, but a closed socket will never be resolved. Should it attempt to reopen the socket each time it's called?

Yes, with the background thread - preferably as soon as the connection fails, not when something is called. Maybe the thread should not terminate and keep monitoring the connection somehow, if that's possible to do at all in Python.

charlesbaynham commented:
Could keep checking the `ping()` method of the `Server`? That's required to exist anyway since otherwise the controller manager kills the server.

charlesbaynham commented:
Actually a better way would be to use the `select` module. See https://docs.python.org/2/howto/sockets.html#non-blocking-sockets or https://stackoverflow.com/questions/17386487/python-detect-when-a-socket-disconnects-for-any-reason

sbourdeauducq commented:
The main issue will be with the concurrent access from the thread.

charlesbaynham commented:
OK, I've reimplemented BestEffortClient as discussed. It now has a continuous loop in a separate thread which monitors the connection and reconnects when required. I've also removed a lot of the repeated logic in BestEffortClient in favour of composition, using a Client. 

I've exposed the status of all the three clients through `is_closed()`, which involved adding `get_rpc_method_list()` to the asyncio client as well (which is probably sensible since both the others have it already). 

I've added tests for the new features. 

### Todo:

* [ ] Release notes

charlesbaynham commented:
If someone is feeling motivated it might be nice to give `AsyncioClient` the same treatment: I think it could use the logic in Client fairly easily, which reduce reduncancy. But it works already, so I think it's fine for now.

sbourdeauducq commented on sipyco/pc_rpc.py:
```diff
@@ -19,10 +19,12 @@
 from operator import itemgetter
 
-from sipyco.monkey_patches import *
+from sipyco.monkey_patches import *  # noqa
```
Those pseudo-comments aren't used anywhere else.

sbourdeauducq commented on sipyco/pc_rpc.py:
```diff
@@ -274,9 +344,14 @@ async def __do_rpc(self, name, args, kwargs):
     def __getattr__(self, name):
         if name not in self.__valid_methods:
-            raise AttributeError
+            raise AttributeError("{} not found".format(name))
```
This should be in a separate PR.

sbourdeauducq commented on sipyco/pc_rpc.py:
```diff
@@ -296,126 +371,135 @@ class BestEffortClient:
+        else:
+            def none_returner(*args, **kwargs):
+                return None
+            return none_returner
```
This can make errors unnecessarily persistent, no?
```python
x = client.foo
while True:
  x()
```

charlesbaynham replied:
Yep, you're right. I've pushed some changes to resolve that by checking for the connection on each call.

charlesbaynham commented on sipyco/pc_rpc.py:
```diff
@@ -296,126 +371,133 @@ class BestEffortClient:
-                raise ValueError
+        with self.__client_lock:
+            return self.__client.is_closed()
 
     def __getattr__(self, name):
-        if name not in self.__valid_methods:
-            raise AttributeError
-        def proxy(*args, **kwargs):
-            return self.__do_rpc(name, args, kwargs)
-        return proxy
+        """
+        If the client is connected, pass this call onto the client. If not,
+        return a method that returns None if the call was valid or if we can't
+        tell. Raise AttributeError otherwise.
+        """
 
-    def get_selected_target(self):
-        raise NotImplementedError
+        if (
+            self.__valid_methods
+            and name not in self.__valid_methods
+            and name not in dir(Client)
+        ):
+            raise AttributeError("{} not found in cached list".format(name))
+        else:
+            # If the client exists, pass the call to the client
+            def tryer(*args, **kwargs):
+                if not self.__client:
+                    return None
 
-    def get_local_host(self):
-        raise NotImplementedError
+                with self.__client_lock:
+                    client_method = getattr(self.__client, name)
+                    try:
+                        return client_method(*args, **kwargs)
+                    except ConnectionError:
+                        self.__mark_invalid()
+                        return None
+            return tryer
```
There's a bug here: if the `Server` returns a `ConnectionError` as the valid, unpacked output of its method this will be misinterpreted as a failed connection. I've added a failing test to highlight this, will have a think as to how to solve.

charlesbaynham replied:
This is resolved in the latest commit: Client now raises `RPCConnectionError` which is a subclass of `ConnectionError` if an RPC connection error occurs, which allows you to tell the difference between errors arising from the RPC code and errors that were unpacked. This is backwards compatible because any it's still a `ConnectionError` so any code which looks for these will still receive them.

charlesbaynham commented:
I notice that there's no release notes in this package. Could add some?

charlesbaynham commented:
@sbourdeauducq I noticed this one is still waiting. Do you think it's ready for merge?

sbourdeauducq commented:
Sorry, I don't currently have time to properly review this. This is a rather large and complex change and I'm too busy with the Zynq port.

charlesbaynham commented:
> Sorry, I don't currently have time to properly review this. This is a rather large and complex change and I'm too busy with the Zynq port.

No worries, just making sure it wasn't forgotten
